### ISSUE TITLE
Clearing group cache - "Can't set headers after they are sent"

### ISSUE THREAD

**Issue by renaudfv** [idealley/feathers-hooks-rediscache#26]
Hi, I created a custom hook on my services for **after**:  _**create**_, _**update**_, _**patch**_ and _**delete**_ in order to clear the group cache made on _**find**_ and _**get**_.

Say my service is called _items_ .
Calling _/cache/clear/group/group-items_ will return Error 500 stating that  "Can't set headers after they are sent".

That is due to the **if (err) ** handler in the router not featuring an else statement for the rest of the callback. _res.status().json_ is then called twice, hence the error.

```javascript 
router.get('/clear/group/:target', function (req, res) {
    client.get('' + req.params.target, function (err, reply) {
      if (err) res.status(500).json({ message: 'something went wrong' });
      var group = reply ? JSON.parse(reply).cache.group : '';
      console.log('TESTTSTTT')
      h.clearGroup(group).then(function (r) {
        res.status(200).json({
          message: 'cache ' + (r ? '' : 'already') + ' cleared for the group key: ' + req.params.target,
          status: r ? 200 : 204
        });
      });
    });
  });
```
Shouldn't the router be something like the following ? 
 
```javascript 
router.get('/clear/group/:target', function (req, res) {
  client.get('' + req.params.target, function (err, reply) {
    if (!!err) {
      res.status(500).json({ message: 'something went wrong' });
    }
    else {
      var group = reply ? JSON.parse(reply).cache.group : '';
      
      h.clearGroup(group).then(function (r) {
        res.status(200).json({
          message: 'cache ' + (r ? '' : 'already') + ' cleared for the group key: ' + req.params.target,
          status: r ? 200 : 204
        });
      }); 
    }
  });   
});
```

Is also think it could be interesting to add hooks to directly clear the actual redis cache (and not only remove it from the hook return object). Having an internal router only makes the feature available through other HTTP request modules.

Thank you,

Renaud

**Comment by idealley** [comment]
Hello, I will check. Sorry for not answering I did not receive the notification.

**Comment by idealley** [comment]
I am sorry but I cannot reproduce. 

In express when you call `res.status()` it returns a response. The `if(err)` should be called only if the there is something wrong with redis. Therefore the `res.status()` should behave like a early return and the rest of the code should not execute. Could you give me more context to reproduce?

Could you also elaborate on your second point, as the item are normally really deleted from redis the method `clearGroup()` really deletes it from redis. 

to clear the cache internally you can easily use redis directly:

```js
const { promisify } = require('util');
const redis = require('redis');
const client = redis.createClient();
const getAsync = promisify(client.get).bind(redis); // if you want to work with promise
const setAsync = promisify(client.set).bind(redis);
...
```

How would you use a hook? 

Would it help to expose certain methods such as `clearGroup()` so that you good just require those when needed?

**Comment by renaudfv** [comment]
Hi,

Having something wrong with Redis does not mean it can't reach the clear group callback.
Calling `res.status().json()` does not exit the scope of the router callback.
You can reproduce the **Can't set headers after they are sent.** issue using only express : 

``` javascript 
const express = require('express');

const app = express();

app.listen(3001);

app.get('/', function(req, res, next) {
  const err = '1234';
  if (!!err) {
    res.status(500).json({message: 'Error'});
  }

  res.status(200).json({data: [1,2,3]});
});
```
You'll get the error, but the server will still catch the aforementioned one.


Regarding the second point, yes exactly. While I could could import Redis itselft and re-implement what is done by your **RedisCache** class, I think exposing it could benefit to users of you module. I also do not think they should need to make HTTP requests to __localhost__'s internal router using additional modules (request, fetch, etc.) to clear the cache (single or group) if the code is already there.
Exposing **RedisCache** could be a first step defenetly, but the calls could also be wrapped in hooks to be used on after for **Create**, **Update**, **Patch** and **Remove** that way a certain service cache could be cleared if resources are modified.

**Comment by renaudfv** [comment]
Closing as the main issue is fixed in PR #29.

Will send a new PR for the enhancement were the behavior handled by the HTTP routes is wrapped by feathers hooks.
